# Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# Importing Dataframes

In [2]:
# Creating a path variable
path = r'C:\Users\Aaron\CareerFoundry\Achievement 6'

In [3]:
# Importing the MLB Players Salaries And Performance dataset
df_mlb = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'salaries_original.csv'), index_col = False)

In [4]:
# Checking the dataframe
df_mlb.head()

index    salary               name  total_value pos  length    years  \
0  20975   6500000  Giancarlo Stanton    325000000  RF      13  2015-27   
1  20430  21943026     Miguel Cabrera    292000000  1B      10  2014-23   
2  20445  22000000     Miguel Cabrera    292000000  1B      10  2014-23   
3  14441  28000000     Alex Rodriguez    275000000  DH      10  2008-17   
4  14447  33000000     Alex Rodriguez    275000000  DH      10  2008-17   

   avg_annual  team  
0     25000000  MIA  
1     29200000  DET  
2     29200000  DET  
3     27500000  NYY  
4     27500000  NYY

### Understanding the Data

In [5]:
df_mlb.shape

(22836, 9)

In [6]:
df_mlb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22836 entries, 0 to 22835
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   index        22836 non-null  int64 
 1   salary       22836 non-null  int64 
 2   name         22836 non-null  object
 3   total_value  22836 non-null  int64 
 4   pos          22836 non-null  object
 5   length       22836 non-null  int64 
 6   years        22836 non-null  object
 7   avg_annual   22836 non-null  int64 
 8   team         22836 non-null  object
dtypes: int64(5), object(4)
memory usage: 1.6+ MB


# Wrangling and Cleaning

#### Dropping 'index' Column

In [7]:
df_mlb = df_mlb.drop(columns = 'index')

In [8]:
# Checking our work
df_mlb.head()

salary               name  total_value pos  length    years  avg_annual   \
0   6500000  Giancarlo Stanton    325000000  RF      13  2015-27     25000000   
1  21943026     Miguel Cabrera    292000000  1B      10  2014-23     29200000   
2  22000000     Miguel Cabrera    292000000  1B      10  2014-23     29200000   
3  28000000     Alex Rodriguez    275000000  DH      10  2008-17     27500000   
4  33000000     Alex Rodriguez    275000000  DH      10  2008-17     27500000   

  team  
0  MIA  
1  DET  
2  DET  
3  NYY  
4  NYY

#### Checking for Missing Values

In [9]:
df_mlb.isnull().sum()

salary         0
name           0
total_value    0
pos            0
length         0
years          0
avg_annual     0
team           0
dtype: int64

#### Checking For Duplicates

In [10]:
df_mlb_dups = df_mlb[df_mlb.duplicated()]

In [11]:
df_mlb_dups

salary               name  total_value pos  length    years  \
6     33000000     Alex Rodriguez    275000000  DH      10  2008-17   
15    24000000      Robinson Cano    240000000  2B      10  2014-23   
26    23000000     Prince Fielder    214000000  1B       9  2012-20   
27    23000000     Prince Fielder    214000000  1B       9  2012-20   
28    24000000     Prince Fielder    214000000  1B       9  2012-20   
...        ...                ...          ...  ..     ...      ...   
3407   2000000       Brendan Ryan      5000000  SS       2  2014-15   
3979   2100000     Ramon Santiago      4200000  2B       2  2012-13   
4352   1900000  Willie Bloomquist      3800000  SS       2  2012-13   
5473   1500000       Gerald Laird      3000000   C       2  2013-14   
5474   1500000        Jeff Mathis      3000000   C       2  2013-14   

      avg_annual  team  
6        27500000  NYY  
15       24000000  SEA  
26       23777778  DET  
27       23777778  TEX  
28       23777778  TEX  
...           ...  ...  
3407      2500000  NYY  
3979      2100000  DET  
4352      1900000  ARI  
5473      1500000  ATL  
5474      1500000  MIA  

[131 rows x 8 columns]

That's a lot of duplicates! We need to drop them.

But first, let's check again how many rows we had to begin with.

In [12]:
df_mlb.shape

(22836, 8)

When we drop the duplicates, we need to have 131 rows less than the original.

In [13]:
df_mlb = df_mlb.drop_duplicates()

In [14]:
df_mlb.shape

(22705, 8)

Perfect! Dropped all duplicate data.

### Creating New 'city' Column

In [15]:
# Create a list of our conditions
conditions = [
    (df_mlb['team'] == 'ARI')
    ,(df_mlb['team'] == 'ATL')
    ,(df_mlb['team'] == 'BAL')
    ,(df_mlb['team'] == 'BOS')
    ,(df_mlb['team'] == 'CHC')
    ,(df_mlb['team'] == 'CWS')
    ,(df_mlb['team'] == 'CIN')
    ,(df_mlb['team'] == 'CLE')
    ,(df_mlb['team'] == 'COL')
    ,(df_mlb['team'] == 'DET')
    ,(df_mlb['team'] == 'HOU')
    ,(df_mlb['team'] == 'KC')
    ,(df_mlb['team'] == 'LAA')
    ,(df_mlb['team'] == 'LAD')
    ,(df_mlb['team'] == 'MIA')
    ,(df_mlb['team'] == 'MIL')
    ,(df_mlb['team'] == 'MIN')
    ,(df_mlb['team'] == 'NYM')
    ,(df_mlb['team'] == 'NYY')
    ,(df_mlb['team'] == 'OAK')
    ,(df_mlb['team'] == 'PHI')
    ,(df_mlb['team'] == 'PIT')
    ,(df_mlb['team'] == 'SD')
    ,(df_mlb['team'] == 'SEA')
    ,(df_mlb['team'] == 'SF')
    ,(df_mlb['team'] == 'STL')
    ,(df_mlb['team'] == 'TB')
    ,(df_mlb['team'] == 'TEX')
    ,(df_mlb['team'] == 'TOR')
    ,(df_mlb['team'] == 'WSH')
]

In [16]:
# Create a List of Values We Want to Assign for Each Condition
# values1 will be used for city
values1 = ['Phoenix'
           ,'Atlanta'
           ,'Baltimore'
           ,'Boston'
           ,'Chicago'
           ,'Chicago'
           ,'Cincinnati'
           ,'Cleveland'
           ,'Denver'
           ,'Detroit'
           ,'Houston'
           ,'Kansas City'
           ,'Los Angeles'
           ,'Los Angeles'
           ,'Miami'
           ,'Milwaukee'
           ,'Minneapolis'
           ,'Queens'
           ,'Bronx'
           ,'Oakland'
           ,'Philadelphia'
           ,'Pittsburgh'
           ,'San Diego'
           ,'Seattle'
           ,'San Francisco'
           ,'St. Louis'
           ,'Tampa Bay'
           ,'Arlington'
           ,'Toronto'
           ,'Washington D.C.'
          ]

In [17]:
# Creating a new column 'city' and use np.select to assing values to it using our lists as arguments
df_mlb['city'] = np.select(conditions, values1)

In [18]:
# Checking our updated df
df_mlb.head(50)

salary               name  total_value pos  length    years  \
0    6500000  Giancarlo Stanton    325000000  RF      13  2015-27   
1   21943026     Miguel Cabrera    292000000  1B      10  2014-23   
2   22000000     Miguel Cabrera    292000000  1B      10  2014-23   
3   28000000     Alex Rodriguez    275000000  DH      10  2008-17   
4   33000000     Alex Rodriguez    275000000  DH      10  2008-17   
5   32000000     Alex Rodriguez    275000000  DH      10  2008-17   
7   30000000     Alex Rodriguez    275000000  DH      10  2008-17   
8   29000000     Alex Rodriguez    275000000  DH      10  2008-17   
9   22000000     Alex Rodriguez    275000000  DH      10  2008-17   
10   3868852     Alex Rodriguez    275000000  DH      10  2008-17   
11  12000000      Albert Pujols    240000000  1B      10  2012-20   
12  24000000      Robinson Cano    240000000  2B      10  2014-23   
13  23000000      Albert Pujols    240000000  1B      10  2012-21   
14  24000000      Albert Pujols    240000000  1B      10  2012-21   
16  16000000      Albert Pujols    240000000  1B      10  2012-20   
17  14000000         Joey Votto    225000000  1B      10  2014-23   
18  12000000         Joey Votto    225000000  1B      10  2014-23   
19  12000000    Clayton Kershaw    215000000   P       7  2014-20   
20  23000000    Clayton Kershaw    215000000  SP       7  2014-20   
21  24000000    Clayton Kershaw    215000000  SP       7  2014-20   
22  16000000    Clayton Kershaw    215000000   P       7  2014-20   
23  23000000     Prince Fielder    214000000  1B       9  2012-20   
24  23000000     Prince Fielder    214000000  1B       9  2012-20   
25  24000000     Prince Fielder    214000000  1B       9  2012-20   
29  17142857       Max Scherzer    210000000  SP       7  2015-21   
30  23000000          Joe Mauer    184000000  1B       8  2011-18   
35  20625000      Mark Teixeira    180000000  1B       8  2009-16   
36  23125000      Mark Teixeira    180000000  1B       8  2009-16   
39  28000000   Justin Verlander    180000000  SP       7  2013-19   
43  20100000   Justin Verlander    180000000  SP       7  2013-19   
44  20000000   Justin Verlander    180000000  SP       7  2013-19   
45  24857142    Felix Hernandez    175000000  SP       7  2013-19   
46  22857142    Felix Hernandez    175000000  SP       7  2013-19   
47  19857143    Felix Hernandez    175000000  SP       7  2013-19   
48  17277777       Buster Posey    167000000   C       9  2013-21   
49  11277777       Buster Posey    167000000   C       9  2013-21   
50   8000000       Buster Posey    167000000   C       9  2013-21   
51  10250000          Matt Kemp    160000000  RF       8  2012-19   
52  21250000          Matt Kemp    160000000  RF       8  2012-19   
53  20250000          Matt Kemp    160000000  RF       8  2012-19   
54  21250000          Matt Kemp    160000000  RF       8  2012-19   
55   8250000    Troy Tulowitzki    157750000  SS      10  2011-20   
56   5500000    Troy Tulowitzki    157750000  SS      10  2011-20   
57  20000000    Troy Tulowitzki    157750000  SS      10  2011-20   
58  16000000    Troy Tulowitzki    157750000  SS      10  2011-20   
59  10000000    Troy Tulowitzki    157750000  SS      10  2011-20   
60  22000000    Masahiro Tanaka    155000000  SP       7  2014-20   
62  20000000         Jon Lester    155000000  SP       6  2015-20   
63  21857142    Adrian Gonzalez    154000000  1B       7  2012-18   
65  21857142    Adrian Gonzalez    154000000  1B       7  2012-18   

    avg_annual  team             city  
0      25000000  MIA            Miami  
1      29200000  DET          Detroit  
2      29200000  DET          Detroit  
3      27500000  NYY            Bronx  
4      27500000  NYY            Bronx  
5      27500000  NYY            Bronx  
7      27500000  NYY            Bronx  
8      27500000  NYY            Bronx  
9      27500000  NYY            Bronx  
10     27500000  NYY            Bronx  
11     24000000  LAA      Los Angeles  
12    

### Creating New 'state' Column

In [19]:
# Create a list of our conditions
conditions = [
    (df_mlb['team'] == 'ARI')
    ,(df_mlb['team'] == 'ATL')
    ,(df_mlb['team'] == 'BAL')
    ,(df_mlb['team'] == 'BOS')
    ,(df_mlb['team'] == 'CHC')
    ,(df_mlb['team'] == 'CWS')
    ,(df_mlb['team'] == 'CIN')
    ,(df_mlb['team'] == 'CLE')
    ,(df_mlb['team'] == 'COL')
    ,(df_mlb['team'] == 'DET')
    ,(df_mlb['team'] == 'HOU')
    ,(df_mlb['team'] == 'KC')
    ,(df_mlb['team'] == 'LAA')
    ,(df_mlb['team'] == 'LAD')
    ,(df_mlb['team'] == 'MIA')
    ,(df_mlb['team'] == 'MIL')
    ,(df_mlb['team'] == 'MIN')
    ,(df_mlb['team'] == 'NYM')
    ,(df_mlb['team'] == 'NYY')
    ,(df_mlb['team'] == 'OAK')
    ,(df_mlb['team'] == 'PHI')
    ,(df_mlb['team'] == 'PIT')
    ,(df_mlb['team'] == 'SD')
    ,(df_mlb['team'] == 'SEA')
    ,(df_mlb['team'] == 'SF')
    ,(df_mlb['team'] == 'STL')
    ,(df_mlb['team'] == 'TB')
    ,(df_mlb['team'] == 'TEX')
    ,(df_mlb['team'] == 'TOR')
    ,(df_mlb['team'] == 'WSH')
]

In [20]:
# Create a List of Values We Want to Assign for Each Condition
# values2 will be used for 'state'
values2 = ['Arizona'
           ,'Georgia'
           ,'Maryland'
           ,'Massachusetts'
           ,'Illinois'
           ,'Illinois'
           ,'Ohio'
           ,'Ohio'
           ,'Colorado'
           ,'Michigan'
           ,'Texas'
           ,'Missouri'
           ,'California'
           ,'California'
           ,'Florida'
           ,'Wisconsin'
           ,'Minnesota'
           ,'New York'
           ,'New York'
           ,'California'
           ,'Pennsylvania'
           ,'Pennsylvania'
           ,'California'
           ,'Washington'
           ,'California'
           ,'Missouri'
           ,'Florida'
           ,'Texas'
           ,'Ontario'
           ,'Maryland'
          ]

In [21]:
# Creating a new column 'state' and use np.select to assign values to it using our lists as arguments
df_mlb['state'] = np.select(conditions, values2)

In [22]:
# Checking our updated df
df_mlb.tail(50)

salary               name  total_value pos  length years  avg_annual   \
22786   62500         Jim Eppard        62500  OF       1  1988        62500   
22787   62500       Jim Sundberg        62500   C       1  1988        62500   
22788   62500         Joey Meyer        62500  1B       1  1988        62500   
22789   62500      John Costello        62500   P       1  1988        62500   
22790   62500        John Dopson        62500   P       1  1988        62500   
22791   62500        John Fishel        62500  OF       1  1988        62500   
22792   62500        John Smoltz        62500   P       1  1988        62500   
22793   62500       Jose Alvarez        62500   P       1  1988        62500   
22794   62500        Jose Cecena        62500   P       1  1988        62500   
22795   62500   Jose J. Bautista        62500   P       1  1988        62500   
22796   62500        Kelly Paris        62500  1B       1  1988        62500   
22797   62500    Ken Griffey Sr.        62500  OF       1  1988        62500   
22798   62500   Larry McWilliams        62500   P       1  1988        62500   
22799   62500      Larry Parrish        62500  1B       1  1988        62500   
22800   62500       Larry Walker        62500  OF       1  1988        62500   
22801   62500        Luis Aquino        62500   P       1  1988        62500   
22802   62500      Mackey Sasser        62500   C       1  1988        62500   
22803   62500         Mario Diaz        62500  SS       1  1988        62500   
22804   62500         Mario Soto        62500   P       1  1988        62500   
22805   62500         Mark Grace        62500  1B       1  1988        62500   
22806   62500     Marvin Freeman        62500   P       1  1988        62500   
22807   62500       Melido Perez        62500   P       1  1988        62500   
22808   62500      Mike Schooler        62500   P       1  1988        62500   
22809   62500  Nelson Santovenia        62500   C       1  1988        62500   
22810   62500      Norm Charlton        62500   P       1  1988        62500   
22811   62500     Oswaldo Peraza        62500   P       1  1988        62500   
22812   62500         Otis Nixon        62500  OF       1  1988        62500   
22813   62500     Ramon Martinez        62500   P       1  1988        62500   
22814   62500        Rick Cerone        62500   C       1  1988        62500   
22815   62500       Ricky Jordan        62500  1B       1  1988        62500   
22816   62500         Rob Dibble        62500   P       1  1988        62500   
22817   62500     Roberto Alomar        62500  2B       1  1988        62500   
22818   62500        Rod Nichols        62500   P       1  1988        62500   
22819   62500      Roger Samuels        62500   P       1  1988        62500   
22820   62500          Ron Jones        62500  OF       1  1988        62500   
22821   62500        Ron Tingley        62500   C       1  1988        62500   
22822   62500        Russ Morman        62500  1B       1  1988        62500   
22823   62500       Shane Turner        62500  3B       1  1988        62500   
22824   62500    Sherman Corbett        62500   P       1  1988        62500   
22825   62500      Sil Campusano        62500  OF       1  1988        62500   
22826   62500      Steve Balboni        62500  1B       1  1988        62500   
22827   62500    Steve Rosenberg        62500   P       1  1988        62500   
22828   62500       Steve Searcy        62500   P       1  1988        62500   
22829   62500        Terry Clark        62500   P       1  1988        62500   
22830   62500       Terry Taylor        62500   P       1  1988        62500   
22831   62500        Thad Bosley        62500  OF       1  1988        62500   
22832   62500          Tim Jones        62500  SS       1  1988        62500   
22833   62500         Todd Burns        62500   P       1  1988        62500   
22834   62500         Tom Prince        62500   C       1  1988        62500   
22835  

### Creating New 'country' Column

In [23]:
# Create a list of our conditions
conditions = [
    (df_mlb['team'] == 'ARI')
    ,(df_mlb['team'] == 'ATL')
    ,(df_mlb['team'] == 'BAL')
    ,(df_mlb['team'] == 'BOS')
    ,(df_mlb['team'] == 'CHC')
    ,(df_mlb['team'] == 'CWS')
    ,(df_mlb['team'] == 'CIN')
    ,(df_mlb['team'] == 'CLE')
    ,(df_mlb['team'] == 'COL')
    ,(df_mlb['team'] == 'DET')
    ,(df_mlb['team'] == 'HOU')
    ,(df_mlb['team'] == 'KC')
    ,(df_mlb['team'] == 'LAA')
    ,(df_mlb['team'] == 'LAD')
    ,(df_mlb['team'] == 'MIA')
    ,(df_mlb['team'] == 'MIL')
    ,(df_mlb['team'] == 'MIN')
    ,(df_mlb['team'] == 'NYM')
    ,(df_mlb['team'] == 'NYY')
    ,(df_mlb['team'] == 'OAK')
    ,(df_mlb['team'] == 'PHI')
    ,(df_mlb['team'] == 'PIT')
    ,(df_mlb['team'] == 'SD')
    ,(df_mlb['team'] == 'SEA')
    ,(df_mlb['team'] == 'SF')
    ,(df_mlb['team'] == 'STL')
    ,(df_mlb['team'] == 'TB')
    ,(df_mlb['team'] == 'TEX')
    ,(df_mlb['team'] == 'TOR')
    ,(df_mlb['team'] == 'WSH')
]

In [24]:
# Create a List of Values We Want to Assign for Each Condition
# values3 will be used for 'country'
values3 = ['USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'USA'
           ,'Canada'
           ,'USA'
          ]

In [25]:
# Creating a new column 'country' and use np.select to assign values to it using our lists as arguments
df_mlb['country'] = np.select(conditions, values3)

In [26]:
# Checking our updated df
df_mlb.tail(50)

salary               name  total_value pos  length years  avg_annual   \
22786   62500         Jim Eppard        62500  OF       1  1988        62500   
22787   62500       Jim Sundberg        62500   C       1  1988        62500   
22788   62500         Joey Meyer        62500  1B       1  1988        62500   
22789   62500      John Costello        62500   P       1  1988        62500   
22790   62500        John Dopson        62500   P       1  1988        62500   
22791   62500        John Fishel        62500  OF       1  1988        62500   
22792   62500        John Smoltz        62500   P       1  1988        62500   
22793   62500       Jose Alvarez        62500   P       1  1988        62500   
22794   62500        Jose Cecena        62500   P       1  1988        62500   
22795   62500   Jose J. Bautista        62500   P       1  1988        62500   
22796   62500        Kelly Paris        62500  1B       1  1988        62500   
22797   62500    Ken Griffey Sr.        62500  OF       1  1988        62500   
22798   62500   Larry McWilliams        62500   P       1  1988        62500   
22799   62500      Larry Parrish        62500  1B       1  1988        62500   
22800   62500       Larry Walker        62500  OF       1  1988        62500   
22801   62500        Luis Aquino        62500   P       1  1988        62500   
22802   62500      Mackey Sasser        62500   C       1  1988        62500   
22803   62500         Mario Diaz        62500  SS       1  1988        62500   
22804   62500         Mario Soto        62500   P       1  1988        62500   
22805   62500         Mark Grace        62500  1B       1  1988        62500   
22806   62500     Marvin Freeman        62500   P       1  1988        62500   
22807   62500       Melido Perez        62500   P       1  1988        62500   
22808   62500      Mike Schooler        62500   P       1  1988        62500   
22809   62500  Nelson Santovenia        62500   C       1  1988        62500   
22810   62500      Norm Charlton        62500   P       1  1988        62500   
22811   62500     Oswaldo Peraza        62500   P       1  1988        62500   
22812   62500         Otis Nixon        62500  OF       1  1988        62500   
22813   62500     Ramon Martinez        62500   P       1  1988        62500   
22814   62500        Rick Cerone        62500   C       1  1988        62500   
22815   62500       Ricky Jordan        62500  1B       1  1988        62500   
22816   62500         Rob Dibble        62500   P       1  1988        62500   
22817   62500     Roberto Alomar        62500  2B       1  1988        62500   
22818   62500        Rod Nichols        62500   P       1  1988        62500   
22819   62500      Roger Samuels        62500   P       1  1988        62500   
22820   62500          Ron Jones        62500  OF       1  1988        62500   
22821   62500        Ron Tingley        62500   C       1  1988        62500   
22822   62500        Russ Morman        62500  1B       1  1988        62500   
22823   62500       Shane Turner        62500  3B       1  1988        62500   
22824   62500    Sherman Corbett        62500   P       1  1988        62500   
22825   62500      Sil Campusano        62500  OF       1  1988        62500   
22826   62500      Steve Balboni        62500  1B       1  1988        62500   
22827   62500    Steve Rosenberg        62500   P       1  1988        62500   
22828   62500       Steve Searcy        62500   P       1  1988        62500   
22829   62500        Terry Clark        62500   P       1  1988        62500   
22830   62500       Terry Taylor        62500   P       1  1988        62500   
22831   62500        Thad Bosley        62500  OF       1  1988        62500   
22832   62500          Tim Jones        62500  SS       1  1988        62500   
22833   62500         Todd Burns        62500   P       1  1988        62500   
22834   62500         Tom Prince        62500   C       1  1988        62500   
22835  

## Understanding the Data

In [27]:
df_mlb.describe()

salary   total_value        length   avg_annual 
count  2.270500e+04  2.270500e+04  22705.000000  2.270500e+04
mean   2.152284e+06  3.708879e+06      1.125875  2.194980e+06
std    3.352784e+06  1.423080e+07      0.751445  3.455491e+06
min    6.250000e+04  6.250000e+04      1.000000  6.250000e+04
25%    3.160000e+05  3.160000e+05      1.000000  3.160000e+05
50%    6.250000e+05  6.350000e+05      1.000000  6.350000e+05
75%    2.600000e+06  2.700000e+06      1.000000  2.633333e+06
max    3.300000e+07  3.250000e+08     13.000000  4.000000e+07

# Exporting Data

In [28]:
df_mlb.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'salaries_cleaned_data.csv'))